### PR TITLE
feat(eslint-plugin): add extension rule `keyword-spacing`

### LINKED ISSUE #1729: [keyword-spacing] Require space before / after `as`
**Repro**
```
const foo = {}as object;
```

```JSON
{
  "rules": {
    "keyword-spacing": "warn",
  }
}
```

**Expected Result**

Warning about a missing space before `as`.

**Actual Result**

No warning.

**Additional Info**

It would be nice to have a `@typescript-eslint/keyword-spacing` rule to cover the typescript keywords.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.23.0` |
| `@typescript-eslint/parser`        | `2.23.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.16.1` |
| `npm`                              | `6.13.4` |

### PR DESCRIPTION
**Pull request by OoDeLally**
Addressing my own issue
fixes #1729

I am totally new to this, so feel free to suggest me improvements.

### CODE REVIEW

File: packages/eslint-plugin/package.json [hunk 1]
```diff
@@ -42,6 +42,7 @@
   },
   "dependencies": {
     "@typescript-eslint/experimental-utils": "2.23.0",
+    "eslint": "^6.8.0",
```
Comment (OoDeLally) [inline packages/eslint-plugin/package.json]: I need `eslint` because I import a list from it
https://github.com/typescript-eslint/typescript-eslint/pull/1739/files#diff-98976b0a1b01edcf278c08cbf1cb4bc8R1
It seems a bit overkill, so I am considering copypasting this list inside my own file. What do you think?

File: packages/eslint-plugin/src/rules/keyword-spacing.ts [hunk 2]
```diff
@@ -0,0 +1,598 @@
+import keywords from 'eslint/lib/rules/utils/keywords';
```
Comment (OoDeLally) [inline packages/eslint-plugin/src/rules/keyword-spacing.ts]: Here I basically took `keyword-spacing.js` from `eslint` and tried to adapt it by watching how it was done for other existing rules.

File: packages/eslint-plugin/src/rules/keyword-spacing.ts [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+//------------------------------------------------------------------------------
+
+const PREV_TOKEN = /^[)\]}>]$/u;
+const NEXT_TOKEN = /^(?:[([{<~!]|\+\+?|--?)$/u;
+const PREV_TOKEN_M = /^[)\]}>*]$/u;
+const NEXT_TOKEN_M = /^[{*]$/u;
+const TEMPLATE_OPEN_PAREN = /\$\{$/u;
+const TEMPLATE_CLOSE_PAREN = /^\}/u;
+const CHECK_TYPE = /^(?:JSXElement|RegularExpression|String|Template)$/u;
+const KEYS = keywords.concat([
```
Comment (OoDeLally) [inline packages/eslint-plugin/src/rules/keyword-spacing.ts]: I could hardcode the list of keyword here instead of importing it from `eslint`.

File: packages/eslint-plugin/src/rules/keyword-spacing.ts [hunk 4]
```diff
@@ -0,0 +1,598 @@
...
+    function checkSpacingForProperty(
+      node: TSESTree.MethodDefinition | TSESTree.Property,
+    ) {
+      if ((node as any).static) {
+        checkSpacingAroundFirstToken(node);
+      }
+      if (
+        node.kind === 'get' ||
+        node.kind === 'set' ||
+        (((node as any).method || node.type === 'MethodDefinition') &&
```
Comment (OoDeLally) [inline packages/eslint-plugin/src/rules/keyword-spacing.ts]: Any way to avoid these ugly `as any`?

File: packages/eslint-plugin/src/rules/keyword-spacing.ts [hunk 5]
```diff
@@ -0,0 +1,598 @@
...
+    function checkSpacingForAwaitExpression(node: TSESTree.AwaitExpression) {
+      checkSpacingBefore(sourceCode.getFirstToken(node)!);
+    }
+
+    /**
+     * Reports `as` keyword of a given node if usage of spacing before
+     * this keyword is invalid.
+     * @param node A node to report.
+     */
+    function checkSpacingForAsExpression(node: TSESTree.TSAsExpression) {
```
Comment (OoDeLally) [inline packages/eslint-plugin/src/rules/keyword-spacing.ts]: Real addition here.

File: packages/eslint-plugin/src/rules/keyword-spacing.ts [hunk 6]
```diff
@@ -0,0 +1,598 @@
...
+      Super: checkSpacingBeforeFirstToken,
+      ThisExpression: checkSpacingBeforeFirstToken,
+      UnaryExpression: checkSpacingBeforeFirstToken,
+      YieldExpression: checkSpacingBeforeFirstToken,
+
+      // Others
+      ImportNamespaceSpecifier: checkSpacingForImportNamespaceSpecifier,
+      MethodDefinition: checkSpacingForProperty,
+      Property: checkSpacingForProperty,
+      TSAsExpression: checkSpacingForAsExpression,
```
Comment (OoDeLally) [inline packages/eslint-plugin/src/rules/keyword-spacing.ts]: Real addition here.

File: packages/eslint-plugin/src/types/eslint.d.ts [hunk 7]
```diff
@@ -0,0 +1,4 @@
+declare module 'eslint/lib/rules/utils/keywords' {
```
Comment (OoDeLally) [inline packages/eslint-plugin/src/types/eslint.d.ts]: Could go away if we toss `eslint` dependency.

File: packages/eslint-plugin/tests/rules/keyword-spacing.test.ts [hunk 8]
```diff
@@ -0,0 +1,4168 @@
...
+      parserOptions: { ecmaVersion: 6, sourceType: 'module' },
+    },
+    {
+      code: 'import *as a from "foo"',
+      options: [override('as', NEITHER)],
+      parserOptions: { ecmaVersion: 6, sourceType: 'module' },
+    },
+
+    //----------------------------------------------------------------------
+    // as (typing)
```
Comment (OoDeLally) [inline packages/eslint-plugin/tests/rules/keyword-spacing.test.ts]: Valid tests here. Not sure if they should be grouped with the normal `as`.

File: packages/eslint-plugin/tests/rules/keyword-spacing.test.ts [hunk 9]
```diff
@@ -0,0 +1,4168 @@
...
+    ';var a = b',
+    { code: '; var a = b', options: [NEITHER] },
+
+    //----------------------------------------------------------------------
+    // void
+    //----------------------------------------------------------------------
+
+    '{} void foo',
+    { code: '{}void foo', options: [NEITHER] },
+    { code: '{} void foo', options: [over
```
Comment (OoDeLally) [inline packages/eslint-plugin/tests/rules/keyword-spacing.test.ts]: Invalidity tests here. Not sure if they should be grouped with the normal `as`.

File: packages/eslint-plugin/src/rules/keyword-spacing.ts (hunk 4 above)
Comment (G-Rath) [reply packages/eslint-plugin/src/rules/keyword-spacing.ts]: You can do checks against the [discriminatory property](https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions) to narrow the types, which will let you access properties without casting :)

For example, in your first `if` condition, you can do:

```
if (
  node.type === AST_NODE_TYPES.MethodDefinition &&
  node.static // 'node' is now type TSESTree.MethodDefintion until the closing } or another guard narrows it
) {
  //
}
```

Generally, it's always better to cast to the exact type you need rather than `any` - and from there you can work backwards to figure out what checks you need to do to narrow the type down without the cast.

When working with AST nodes, typically it's about checking their `type`, but with other less uniform objects you might need to write a custom type guard.

File: packages/eslint-plugin/src/rules/keyword-spacing.ts (hunk 6 above)
Comment (G-Rath) [reply packages/eslint-plugin/src/rules/keyword-spacing.ts]: If this is the only addition (along with it's function), you hopefully should be able to just re-use the base rule, and merge your addition in with the rule.

Check out [`@typescript-eslint/quotes`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/quotes.ts) & [`@typescript-eslint/semi`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/semi.ts) for examples of how to do this.

(You still might have to duplicate some code in, but it should at least remove the need to duplicate this return).

File: packages/eslint-plugin/tests/rules/keyword-spacing.test.ts (hunk 9 above)
Comment (G-Rath) [inline packages/eslint-plugin/tests/rules/keyword-spacing.test.ts]: This shouldn't be needed - if it is, that's a sign that there's a type that needs to be adjusted somewhere.

(My first guess would be the return types of the helper methods)

Comment (G-Rath) [reply packages/eslint-plugin/tests/rules/keyword-spacing.test.ts]: :/ this was meant to be on the [`as TSESLint.InvalidTestCase<MessageIds, Options>[],`](https://github.com/typescript-eslint/typescript-eslint/pull/1739/files#diff-9ec4a33b0114a94ee107a0f20c938ebeR4167).

File: packages/eslint-plugin/src/types/eslint.d.ts (hunk 7 above)
Comment (OoDeLally) [reply packages/eslint-plugin/src/types/eslint.d.ts]: Can you please advise whether we can avoid this file.
The fact that I've seen none if this kind in the repo makes me think that there is another way.

File: packages/eslint-plugin/src/rules/keyword-spacing.ts (hunk 6 above)
Comment (OoDeLally) [reply packages/eslint-plugin/src/rules/keyword-spacing.ts]: Nice! This simplifies a lot. However I still had to redefine most helper functions, since they are not exported by the eslint rule.

File: packages/eslint-plugin/src/rules/keyword-spacing.ts (hunk 4 above)
Comment (OoDeLally) [reply packages/eslint-plugin/src/rules/keyword-spacing.ts]: Got it. Although the problem doesn't exist anymore since I import the base rule.

File: packages/eslint-plugin/src/types/eslint.d.ts (hunk 7 above)
Comment (OoDeLally) [reply packages/eslint-plugin/src/types/eslint.d.ts]: Moved to `typings/eslint-keyword.d.ts`.

**Comment by OoDeLally** [comment]
@G-Rath Ok, thanks for your comments. I'm now importing the base rule, it simplifies a bit.

File: packages/eslint-plugin/typings/eslint-rules.d.ts [hunk 10]
```diff
@@ -142,6 +142,69 @@ declare module 'eslint/lib/rules/indent' {
...
 
+declare module 'eslint/lib/rules/keyword-spacing' {
+  import { TSESLint, TSESTree } from '@typescript-eslint/experimental-utils';
+
+  const rule: TSESLint.RuleModule<
+    'expectedBefore' | 'expectedAfter' | 'unexpectedBefore' | 'unexpectedAfter',
+    [],
+    {
+      // Statements
+      DebuggerStatement: (node: TSESTree.DebuggerStatement) => void;
```
Comment (G-Rath) [inline packages/eslint-plugin/typings/eslint-rules.d.ts]: You could type these using `TSESLint.RuleFunction<T>` to tie them back to the single type, but that might be overkill.

It's probably worth making these methods however, to be consistent with the other types.

File: packages/eslint-plugin/src/rules/keyword-spacing.ts (hunk 6 above)
Comment (G-Rath) [reply packages/eslint-plugin/src/rules/keyword-spacing.ts]: Awesome! While I expected you'd have to duplicate some of the helpers, it's annoying that there isn't any node that calls the methods you want that we could trick into doing the work for us, but oh well it's still a nice simplification 🤷‍♂

File: packages/eslint-plugin/src/rules/keyword-spacing.ts (hunk 3 above)
Comment (G-Rath) [reply packages/eslint-plugin/src/rules/keyword-spacing.ts]: Importing them is completely fine :)

File: packages/eslint-plugin/src/rules/keyword-spacing.ts [hunk 11]
```diff
@@ -0,0 +1,272 @@
+import baseRule from 'eslint/lib/rules/keyword-spacing';
+import { TSESTree } from '@typescript-eslint/experimental-utils';
+import { isTokenOnSameLine } from '../util/astUtils';
+import keywords from 'eslint/lib/rules/utils/keywords';
+
+import * as util from '../util';
+
+export type Option = Partial<{
```
Comment (G-Rath) [inline packages/eslint-plugin/src/rules/keyword-spacing.ts]: This should be `Options` (it's what CI is currently failing on, since it's referenced further down)

**Comment by OoDeLally** [comment]
I'm now re-declaring the list of keywords both in typing and in string array.

**Comment by OoDeLally** [comment]
What is the deal with `Semantic Pull Request`? What is wrong with my title?

**Comment by OoDeLally** [comment]
How is that `codecov` thingy actionable? Anything I should do about it?

File: packages/eslint-plugin/src/rules/keyword-spacing.ts [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+} from '@typescript-eslint/experimental-utils';
+import { isTokenOnSameLine, KEYWORDS } from '../util/astUtils';
+import * as util from '../util';
+
+export type Option = Partial<{
+  before: boolean;
+  after: boolean;
+}>;
+export type OverrideOptions = Partial<Record<Keyword, Option>>;
+export type RootOption = Option & { overrides?: OverrideOptions };
```
Comment (OoDeLally) [inline packages/eslint-plugin/src/rules/keyword-spacing.ts]: How can I avoid declaring these types twice?
I cannot declare them in the `eslint-rule.d.ts` because the `rule` object already occupies the whole module.
And I cannot import them from here to `eslint-rule.d.ts` either...
Any good clean solution?

File: packages/eslint-plugin/typings/eslint-keywords.d.ts [hunk 13]
```diff
@@ -0,0 +1,69 @@
+declare type Keyword =
```
Comment (OoDeLally) [inline packages/eslint-plugin/typings/eslint-keywords.d.ts]: It would be nice to generate this type from the `KEYWORDS` string array in `astUtils.ts`, just like it done in https://stackoverflow.com/a/54061487/1541141
Unfortunately, I am not able to declare a value in a `d.ts` file.
Any way to make this nice?

**Comment by bradzacher** [comment]
> What is the deal with Semantic Pull Request? What is wrong with my title?

You were missing a `:` after your `)`.

> How is that codecov thingy actionable? Anything I should do about it?

If you click through to the details on the codecov site, it'll show you what branches of your code have not been hit by tests.
i.e. clicking on the `keyword-spacing.ts` file in the codecov comment takes you here: https://codecov.io/gh/typescript-eslint/typescript-eslint/pull/1739/diff?src=pr&el=tree#diff-cGFja2FnZXMvZXNsaW50LXBsdWdpbi9zcmMvcnVsZXMva2V5d29yZC1zcGFjaW5nLnRz

Scrolling through that report, green means all good, yellow means one of the branches was missed, and red means all of the branches were missed.
I.e. looking at the report, theres missed lines for reporting `expectedAfter` and `unexpectedAfter`.

File: packages/eslint-plugin/src/util/astUtils.ts [hunk 14]
```diff
@@ -239,6 +239,77 @@ function isMemberOrOptionalMemberExpression(
   );
 }
 
+const KEYWORDS = [
```
Comment (G-Rath) [inline packages/eslint-plugin/src/util/astUtils.ts]: I'm not really sure of the value of this, as having it here means we have to remember to update items if the list ever changes, and that we're now shipping code that is already available elsewhere (since we can just import it from eslint).

File: packages/eslint-plugin/typings/eslint-keywords.d.ts (hunk 13 above)
Comment (G-Rath) [reply packages/eslint-plugin/typings/eslint-keywords.d.ts]: This file is only needed if you're importing from `eslint-keywords` - since you've now pulled it into `astUtils`, you don't need this :)

File: packages/eslint-plugin/src/util/astUtils.ts (hunk 14 above)
Comment (OoDeLally) [reply packages/eslint-plugin/src/util/astUtils.ts]: The list is also hardcoded in the original `keyword-spacing.js`
https://github.com/eslint/eslint/blob/master/lib/rules/keyword-spacing.js
Since the list can only be imported partially, we still have to hardcode some of it, and we may have to change it if a new keyword is introduced.
What do you think?

For now I will import what I can from the list eslint exports.

Comment (G-Rath) [reply packages/eslint-plugin/src/util/astUtils.ts]: Ah yes I see what you mean - that is a pain.

I think importing + concating whats missing is the best way to go, since it's a free deduplication that we'll not have to think much about after it's committed so might as well take it.

I don't think we need to worry about the deduplication check they've got in the rule however, at least for now.

File: packages/eslint-plugin/typings/eslint-keywords.d.ts (hunk 13 above)
Comment (OoDeLally) [reply packages/eslint-plugin/typings/eslint-keywords.d.ts]: The issue here is that I need to know all the possible values of `Keyword` in order to build my `Options` type. However, if I import the list of keywords from `eslint`, the list is only known at runtime.

File: packages/eslint-plugin/src/util/astUtils.ts (hunk 14 above)
Comment (OoDeLally) [reply packages/eslint-plugin/src/util/astUtils.ts]: About this deduplication thingy... It was there in the original file... Maybe for a  reason. It is documented nowhere why a given keyword made it to the offical list, and another doesn't.
This leads me to think that we anticipate the possibility that the `eslint` guys may tamper with that list. In this case, I would favor crashing early, even if it adds a bit of complexity.

File: packages/eslint-plugin/typings/eslint-keywords.d.ts (hunk 13 above)
Comment (G-Rath) [reply packages/eslint-plugin/typings/eslint-keywords.d.ts]: Yeah this is an interesting one.

I think for now the best way to handle this is to type it as `string`, and look at where it breaks - I've just realised that we're going to have to be very careful about this rule anyway due to how the base rule works.

I'll put the details in another comment on the more-relevant lines :)

File: packages/eslint-plugin/src/rules/keyword-spacing.ts [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+  meta: {
+    type: 'layout',
+    docs: {
+      description: 'Enforce consistent spacing before and after keywords',
+      category: 'Stylistic Issues',
+      recommended: false,
+      extendsBaseRule: true,
+    },
+    fixable: 'whitespace',
+    schema: baseRule.meta.schema,
```
Comment (G-Rath) [inline packages/eslint-plugin/src/rules/keyword-spacing.ts]: Looking over the base rule, I realised it uses the `KEYWORDS` array to build the schema, which means that if we were to add our own keywords this rule would error unless we pulled the schema in ourselves (which wouldn't be the end of the world).

Given that, I think it would be best to have the `KEYWORDS` const in here instead of `astUtils`, to make it internal for just this rule - that way it's less visible and so less likely to be added to by people believing it's a standard utility constant.

This also impacts the typings: while having types as accurate as possible is nice, in this case it doesn't matter as much since we're extending a base rule whos internals we can't reach properly anyway, so the effort required to ensure the types are completely sound is probably going to be far more than the value especially given that we ideally want to make as little changes to this rule as possible, and that it's the schema that powers the validation & such.

In light of this, I personally would say that we don't *need* to copy over the deduplication checking function, since our `KEYWORDS` const should never be different from the one in base rules, and the base rule would error anyway if it changes, but I'm not against it because as you say it's a small number of lines.

------

The tl;dr here is that I think we should definitely keep the `KEYWORDS` constant internal to this rule given how it's used by the base rule.

Comment (OoDeLally) [reply packages/eslint-plugin/src/rules/keyword-spacing.ts]: Ok, you got it. So I'll drop the `Keyword[]` type and stick with `string[]`. It will make things simpler.

**Comment by OoDeLally** [comment]
Would you know a way to avoid having to declare the option typing twice (in `eslint-rule.d.ts` and in `keyword-spacing.ts`) ?
I see no way of declaring it at one place and import them from the other place. Tried many ways, without success.

File: packages/eslint-plugin/src/rules/keyword-spacing.ts (hunk 15 above)
Comment (G-Rath) [reply packages/eslint-plugin/src/rules/keyword-spacing.ts]: Sounds good. After posting this I realised there are some other keywords that we want to add: 

* `abstract`,
* `asserts`,
* `declare`,
* `extends` (for interfaces),
* `implements`,
* `interface`,
* `infer`,
* `keyof`,
* `module`,
* `namespace`,
* `private`,
* `protected`,
* `public`,
* `readonly`
* `type`,
* `typeof` (this is handled by a different AST node type, and so technically a different operator from the runtime version), 

So we'll want to extend the schema at some point, but I think for now we should just focus on getting the initial rule landed via this PR, as adding new words will have a small diff and so be very easy to do via another PR :)

File: packages/eslint-plugin/src/rules/keyword-spacing.ts [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+(function(): void {
+  KEYWORDS.sort();
+  for (let i = 1; i < KEYWORDS.length; ++i) {
+    if (KEYWORDS[i] === KEYWORDS[i - 1]) {
+      throw new Error(
+        `Duplication was found in the keyword list: ${KEYWORDS[i]}`,
+      );
+    }
+  }
+})();
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/keyword-spacing.ts:49]: could just use a set

```suggestion
const KEYWORDS = new Set(ESLINT_UTILS_KEYWORDS.concat([
  // Eslint Utils does no provide all keywords, so we complete the list here.
  'as',
  'async',
  'await',
  'from',
  'get',
  'let',
  'of',
  'set',
  'yield',
]));
```

File: packages/eslint-plugin/typings/eslint-rules.d.ts [hunk 17]
```diff
@@ -142,6 +142,79 @@ declare module 'eslint/lib/rules/indent' {
...
+  import { TSESLint, TSESTree } from '@typescript-eslint/experimental-utils';
+  import { RuleFunction } from '@typescript-eslint/experimental-utils/dist/ts-eslint';
+
+  type Option = Partial<{
+    before: boolean;
+    after: boolean;
+  }>;
+  type OverrideOptions = Partial<Record<string, Option>>;
+  type RootOption = Option & { overrides?: OverrideOptions };
+  type Options = [RootOption];
```
Comment (bradzacher) [inline packages/eslint-plugin/typings/eslint-rules.d.ts:155]: Could simplify this down. There's not enough duplication to bother deduping.

```suggestion
  type Options = [
  	{
	  before?: boolean;
	  after?: boolean;
	  overrides?: Record<string, {
	    before?: boolean;
	    after?: boolean;
	  }>;
	}
  ];
```

File: packages/eslint-plugin/src/rules/keyword-spacing.ts [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+} from '@typescript-eslint/experimental-utils';
+import { isTokenOnSameLine } from '../util/astUtils';
+import * as util from '../util';
+
+export type Option = Partial<{
+  before: boolean;
+  after: boolean;
+}>;
+export type OverrideOptions = Partial<Record<string, Option>>;
+export type RootOption = Option & { overrides?: OverrideOptions };
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/keyword-spacing.ts:15]: you shouldn't need to redefine these - just the `Options` type below should be enough
```suggestion
```

File: packages/eslint-plugin/src/rules/keyword-spacing.ts (hunk 6 above)
Comment (bradzacher) [reply packages/eslint-plugin/src/rules/keyword-spacing.ts:290]: note that there is already the keyword `as` registered in the rule due to `import { x as y } from 'foo'`.
Looking at the rule's source code, you might be able to get away with just this fudge code for this extension:

```ts
return {
  ...baseRules,
  TSAsExpression(node): void {
    baseRules.ImportNamespaceSpecifier(node as any);
  },
};
```

**Comment by OoDeLally** [comment]
> Looking at the rule's source code, you might be able to get away with just this fudge code for this extension:
This seems to fail every test with:
```
  ● keyword-spacing › invalid › const foo = {}as {}

    TypeError: Cannot read property 'before' of undefined
    Occurred while linting file.ts:1

      at checkSpacingBefore (../../node_modules/eslint/lib/rules/keyword-spacing.js:268:21)
      at Object.checkSpacingForImportNamespaceSpecifier [as ImportNamespaceSpecifier] (../../node_modules/eslint/lib/rules/keyword-spacing.js:466:13)
      at TSAsExpression (src/rules/keyword-spacing.ts:3054:19)
```
The error is in `lib/rules/keyword-spacing.js` at 
```
        function checkSpacingBefore(token, pattern) {
            checkMethodMap[token.value].before(token, pattern || PREV_TOKEN);
        }
```

I don't really understand what is going on, so I wouldn't risk replacing it.

If you find a way to make it work, I'll be happy to simplify it.

**Comment by OoDeLally** [comment]
I get many errors from `Code style and lint`:

```
##[error]   88:13  error  This snippet should be formatted correctly. Use the fixer to format the code            
```

about my tests
```ts
    {
      code: 'const foo = {} as {}', // < HERE
      parserOptions: { ecmaVersion: 6, sourceType: 'module' },
    },
```

However I don't know what is wrong, and neither `npm run lint` nor `npm run format` report any problem.

**Comment by G-Rath** [comment]
> However I don't know what is wrong, and neither npm run lint nor npm run format report any problem.

Have you tried `npm run lint -- --fix`?

The formatting issue is that it's not got a semi-colon :)

**Comment by bradzacher** [comment]
Okay, I figured the minimal set of code that is required for this rule.
It's a very hacky, but it saves us from duplicating any code from the base ESLint rule, which I think is a huge win for such a minor change.

Unfortunately, you can't just use `ImportNamespaceSpecifier` because when that check was implemented (4 YEARS AGO!!!), [they used `getFirstToken(node, 1)`](https://github.com/eslint/eslint/blob/84f71de0e686e0fe37b83d6728ce1825caaa44fb/lib/rules/keyword-spacing.js#L460), which means "get the first token, skipping 1 token". So for code like `{} as {}`, the tokens are: `[{, }, as, {, }]`, i.e the first token skipping 1 is `}`.

We could ofc submit a fix to the underlying rule to change it to `getFirstToken(node, token => token.value === "as")`, but then that would hard tie this extension rule to the (yet unreleased) ESLint v7, which is obviously a bad thing.

```ts
import { AST_TOKEN_TYPES } from '@typescript-eslint/experimental-utils';
import baseRule from 'eslint/lib/rules/keyword-spacing';
import * as util from '../util';

export type Options = util.InferOptionsTypeFromRule<typeof baseRule>;
export type MessageIds = util.InferMessageIdsTypeFromRule<typeof baseRule>;

export default util.createRule<Options, MessageIds>({
  name: 'keyword-spacing',
  meta: {
    type: 'layout',
    docs: {
      description: 'Enforce consistent spacing before and after keywords',
      category: 'Stylistic Issues',
      recommended: false,
      extendsBaseRule: true,
    },
    fixable: 'whitespace',
    schema: baseRule.meta.schema,
    messages: baseRule.meta.messages,
  },
  defaultOptions: [{}],
  create(context) {
    const sourceCode = context.getSourceCode();
    const baseRules = baseRule.create(context);

    return {
      ...baseRules,
      TSAsExpression(node): void {
        const asToken = util.nullThrows(
          sourceCode.getTokenAfter(
            node.expression,
            token => token.value === 'as',
          ),
          util.NullThrowsReasons.MissingToken('as', node.type),
        );
        const oldTokenType = asToken.type;
        // as is a contextual keyword, so it's always reported as an Identifier
        // the rule looks for keyword tokens, so we temporarily override it
        // we mutate it at the token level because the rule calls sourceCode.getFirstToken,
        // so mutating a copy would not change the underlying copy returned by that method
        asToken.type = AST_TOKEN_TYPES.Keyword;

        // use this selector just because it is just a call to `checkSpacingAroundFirstToken`
        baseRules.DebuggerStatement(asToken as never);

        // make sure to reset the type afterward so we don't permanently mutate the AST
        asToken.type = oldTokenType;
      },
    };
  },
});
```

**Comment by bradzacher** [comment]
> and neither npm run lint nor npm run format report any problem.

Try running `npm run build` in the root to rebuild the internal lint rules.

Note - you should add this comment block to the very top of your test file to disable the prettier formatting
```ts
/* eslint-disable eslint-comments/no-use */
// this rule tests the spacing, which prettier will want to fix and break the tests
/* eslint "@typescript-eslint/internal/plugin-test-formatting": ["error", { formatWithPrettier: false }] */
/* eslint-enable eslint-comments/no-use */
```

**Comment by OoDeLally** [comment]
> Okay, I figured the minimal set of code that is required for this rule.

Nice, well done!

> Try running npm run build in the root to rebuild the internal lint rules.

I did, without any change.

> Note - you should add this comment block to the very top of your test file to disable the prettier formatting

Now the `Code style and lint` succeeded, but my local `yarn lint` fails:
```
/path/to/typescript-eslint/packages/eslint-plugin/tests/rules/keyword-spacing.test.ts
  3:1  error  Definition for rule '@typescript-eslint/internal/plugin-test-formatting' was not found  @typescript-eslint/internal/plugin-test-formatting
```
Is there a missing dep I forgot to install?


Now these tests are stuck
image
